File: Claude/Intermediárias/longest_palindrome_substring/longest_palindrome_substring.py
def longest_palindrome(s):
    # Preprocess the string
    T = '#'.join('^{}$'.format(s))
    n = len(T)
    P = [0] * n
    C = R = 0

    for i in range(1, n-1):
        P[i] = (R > i) and min(R - i, P[2*C - i])
        # Attempt to expand palindrome centered at i
        while T[i + 1 + P[i]] == T[i - 1 - P[i]]:
            P[i] += 1

        # If palindrome centered at i expands past R,
        # adjust center based on expanded palindrome.
        if i + P[i] > R:
            C, R = i, i + P[i]

    # Find the maximum element in P
    maxLen, centerIndex = max((n, i) for i, n in enumerate(P))
    return s[(centerIndex  - maxLen)//2: (centerIndex  + maxLen)//2]

# Test the function
s = "babad"
print(longest_palindrome(s))  # Output: "bab" or "aba"

File: Gemini/Intermediárias/longest_palindrome_substring/longest_palindrome_substring.java
public class LongestPalindromicSubstring {

    public String longestPalindrome(String s) {
        if (s == null || s.length() < 1) return "";
        int start = 0, end = 0;
        String t = preProcess(s);
        int len = t.length();
        int[] p = new int[len];

        for (int i = 1; i < len - 1; i++) {
            int mirror = 2 * center - i;
            p[i] = (i < end) ? Math.min(end - i, p[mirror]) : 0;

            while (t.charAt(i + 1 + p[i]) == t.charAt(i - 1 - p[i])) p[i]++;

            if (i + p[i] > end) {
                center = i;
                end = i + p[i];
            }
        }

        int maxLen = 0;
        for (int i = 1; i < len - 1; i++) {
            if (p[i] > maxLen) {
                start = (i - 1 - p[i]) / 2;
                maxLen = p[i];
            }
        }
        return s.substring(start, start + maxLen);
    }

    private String preProcess(String s) {
        int n = s.length();
        if (n == 0) return "^$";
        String ret = "^";
        for (int i = 0; i < n; i++) ret += "#" + s.charAt(i);
        ret += "#$";
        return ret;
    }
}
